File: src/R2ProtocolTest.js
const fs = require("fs");
const r2i = require("./R2Protocol.js");

const action = process.argv[2];
const input = fs.readFileSync(process.argv[3]);
let data;
if (action === "decode") {
    const params = r2i.decode(input);
    if (params) {
        data = [
            params.start ? "start" : "",
            "source = " + (params.source || ""),
            "destination = " + (params.destination || ""),
            "id = " + (params.id || ""),
            "data = " + (params.data || ""),
            "checksum = " + (params.checksum || ""),
            params.end ? "end" : ""
        ].join("\n");
    }
    else {
        data = "invalid";
    }
}
else if (action === "encode") {
    const lines = input.toString().split("\n");
    const params = {};
    for (const line of lines) {
        const eq = line.indexOf("=");
        if (eq >= 0) {
            params[line.substring(0, eq).trim()] = line.substring(eq + 1).trim();
        }
        else {
            params[line.trim()] = true;
        }
    }
    if (params["start"] && params["source"] && params["destination"] && params["data"] && params["checksum"] && params["end"]) {
        data = r2i.encode(params);
        // TODO compare checksum
    }
    else {
        data = "invalid";
    }
}
fs.writeFileSync(process.argv[4], data);

File: src/R2ProtocolTest.py
import sys
import R2Protocol as r2p

if __name__=="__main__":
    if sys.argv[1] == "decode":
        f = open(sys.argv[2], 'rb')
        data = f.read()
        f.close()

        f = open(sys.argv[3], 'wb')
        params = r2p.decode(data)
        data = b""
        if params is not None:
            data += b"start\n"
            data += b"source = " + params["source"] + b"\n"
            data += b"destination = " + params["destination"] + b"\n"
            data += b"id = " + params["id"] + b"\n"
            data += b"data = " + params["data"] + b"\n"
            data += b"checksum = " + params["checksum"] + b"\n"
            data += b"end"
        else:
            data += b"invalid"

        f.write(data)
        f.close()

    elif sys.argv[1] == "encode":
        f = open(sys.argv[2], 'rb')
        params = {}
        for line in f.readlines():
            eq = line.find(b"=")
            if eq >= 0:
                key = line[:eq].strip().decode("ascii")
                value = line[eq+1:].strip()
                params[key] = value
        
        f.close()

        if "source" in params and "destination" in params and "id" in params and "data" in params:
            data = r2p.encode(params["source"], params["destination"], params["id"], params["data"])
        else:
            data = b"invalid"
        
        f = open(sys.argv[3], 'wb')
        f.write(data)
        f.close()
